### ISSUE TITLE
SFT pipeline freezes random masking before training?

### ISSUE THREAD

**Issue by gaoxin492** [Gen-Verse/dLLM-RL#31]
Hi, thanks for sharing the repo.
But I noticed a critical design issue in your SFT pipeline: `prepare_inputs_and_labels_for_text()` is called before dataset creation, so each sample gets a fixed random mask (ratio + positions) that never changes during training.

Even if you `set mask_times_per_sample > 1`, it only produces a few pre-generated noisy variants — the dataset is still deterministic. (In fact, this parameter seems better interpreted as increasing the number of Monte Carlo samples rather than genuinely adding stochasticity.)
After a few epochs, the model simply overfits these fixed noise patterns, and performance drops.

Diffusion-style training should resample noise per batch, not freeze it in preprocessing. Why not follow most diffusion or masked-LM repos and apply random noise after each batch is sampled? That would keep the data stochastic and prevent overfitting.

**Comment by WilfredGuan** [comment]
> What do you mean by "prepare_inputs_and_labels_for_text() is called before dataset creation"? This function needs the input of dataset (see 1, which is from sft.llada.py), and for each random masking time, we randomly sample different masking (see 2), so no deterministic exists.

Actually, I have a similar question. I think this issue means that if you're training the model with multiple epochs, each randomly masked sample would be trained on multiple times with the same mask pattern.

For example:

```
Epoch 1, sample 1: The cat MASK on the mat.
Epoch 2, sample 1: The cat MASK on the mat.
```

However, it seems that the following case is preferred:

```
Epoch 1, sample 1 (version 1): The cat MASK on the mat.
Epoch 1, sample 1 (version 2): The MASK MASK on MASK mat.
```
